src/FtpExtractor.php

```diff
@@ -172,24 +177,83 @@ private function download(): int
         uasort($this->filesToDownload, $cbTimestampSort);
 
-        $fs = new Filesystem();
         $downloadedFiles = 0;
         foreach ($this->filesToDownload as $file) {
-            $file[self::FILE_DESTINATION_KEY] = ColumnNameSanitizer::toAscii($file[self::FILE_DESTINATION_KEY]);
-
-            $this->logger->info(sprintf("Downloading file %s", $file[self::FILE_SOURCE_KEY]));
-
-            try {
-                $fs->dumpFile(
-                    $file[self::FILE_DESTINATION_KEY],
-                    $this->ftpFilesystem->read($file[self::FILE_SOURCE_KEY])
-                );
-            } catch (\Throwable $e) {
-                ExceptionConverter::handleDownloadException($e);
-            }
-            $this->registry->updateOutputState($file[self::FILE_SOURCE_KEY], $file[self::FILE_TIMESTAMP_KEY]);
+            $this->downloadFile($file);
             $downloadedFiles++;
         }
         return $downloadedFiles;
     }
+
+    private function downloadFile(array $file): void
+    {
+        $file[self::FILE_DESTINATION_KEY] = ColumnNameSanitizer::toAscii($file[self::FILE_DESTINATION_KEY]);
+
+        $this->logger->info(sprintf("Downloading file %s", $file[self::FILE_SOURCE_KEY]));
+
+        $localPath = $file[self::FILE_DESTINATION_KEY];
+        $ftpPath = $file[self::FILE_SOURCE_KEY];
+
+        try {
+            $this->createRetryProxy()->call(function () use ($localPath, $ftpPath): void {
+                $ftpSize = $this->ftpFilesystem->getSize($ftpPath);
+                $this->fs->dumpFile($localPath, $this->ftpFilesystem->read($ftpPath));
+                $localSize = filesize($localPath);
+                $this->checkFileSize($localPath, $ftpPath, $localSize, $ftpSize);
+            });
+        } catch (\Throwable $e) {
+            ExceptionConverter::handleDownloadException($e);
+        }
+        $this->registry->updateOutputState($file[self::FILE_SOURCE_KEY], $file[self::FILE_TIMESTAMP_KEY]);
+    }
+
+    private function createRetryProxy(): RetryProxy
+    {
+        return new RetryProxy(
+            new SimpleRetryPolicy(self::CONNECTION_RETRIES),
+            new ExponentialBackOffPolicy(self::RETRY_BACKOFF),
+            $this->logger
+        );
+    }
+
+    /**
+     * @param string $localPath
+     * @param string $ftpPath
+     * @param int|false $localSize
+     * @param int|false $ftpSize
+     * @throws ApplicationException
+     */
+    private function checkFileSize(string $localPath, string $ftpPath, $localSize, $ftpSize): void
+    {
+        if (!is_int($localSize)) {
+            throw new ApplicationException(
+                sprintf('Cannot get size of the local file "%s".', $localPath)
+            );
+        }
+
+        if (!is_int($ftpSize)) {
+            throw new ApplicationException(
+                sprintf('Cannot get size of the FTP file "%s".', $ftpPath)
+            );
+        }
+
+        if ($ftpSize !== $localSize) {
+            throw new UserException(sprintf(
```

Comment:
Ak nesedi lokalna a FTP velkost suboru tak sa vyhodi chyba a spravi retry. Neda sa na to napisat test, lebo neviem prinutit FTP server neposlat cely subor. No da sa manualne pridat nad podmienku `$ftpSize+=1;` a spustit testy, ... skusal som, vyhodi sa chyba a spravi retry.